### PR TITLE
66360 - Automation of Kms Key Rotation

### PR DESCRIPTION
**Pull request by ryan-mcneil**
## Summary

- Our CMK (Customer Managed Key) in KMS/AWS is automatically rotated annually, and best practice suggests we should be using that key to rotate each encrypted records' `encrypted_kms_key` as well.
- It's been nearly 2 years since we've re-encrypted our data with the current CMK.
- This PR creates a Cron job, that initiates a job that in turn creates Sidekiq::Batch of workers to rotate chunks of records. Upon finishing a batch, the job runs again to grab the next batch of records that need to be updated
- we patch the `has_kms_key` method of the `kms_encrypted` gem to implicitly apply the correct version (determined based on the date that the CMK is rotated), and provides an enumerated list of previous versions, that we can pick up all of the various versions we currently have in the db

## Related issue(s)
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/66360
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/48031

### Subsequent PRs to be merged after this one
- https://github.com/department-of-veterans-affairs/vets-api/pull/14039
- https://github.com/department-of-veterans-affairs/vets-api/pull/14041

## Testing done

- Created various records locally, confirming they were created with the expected version (`v2022`)
- When running the job, I adjusted the date in `kms_encrypted_patch.rb` to something before "today", so that it would pick them up in the query and rotate them as expect.
- I would try different batch and job sizes, based on the number of records created, to confirm that the batch_initiator would properly run upon completing the first batch of jobs.

## What areas of the site does it impact?
* This will update all records that are encrypted using `kms_encrypted`. We've patched the gem in a way that future devs won't need to have knowledge of the change when creating new encrypted models

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

File: app/models/application_record.rb
Comment (ericboehs) [inline app/models/application_record.rb:29]: It does update attributes, just not the updated at ones.

File: app/workers/kms_key_rotation/batch_initiator_job.rb
Comment (ericboehs) [inline app/workers/kms_key_rotation/batch_initiator_job.rb:33]: I'm not sure we should be rescuing all errors here. 😬

File: config/initializers/kms_encrypted_patch.rb
Comment (ericboehs) [inline config/initializers/kms_encrypted_patch.rb:5]: Wouldn't these need the same args? 😕

File: app/workers/kms_key_rotation/batch_initiator_job.rb
Comment (ryan-mcneil) [reply app/workers/kms_key_rotation/batch_initiator_job.rb:33]: I was on the fence. I was thinking that If it threw an error for a one-off issue, then it would be nice to rescue it so that the job would keep going... However I guess if it never pushed the batch, it would never complete, and wouldn't continue anyway. So yeah, I agree.

File: config/initializers/kms_encrypted_patch.rb
Comment (ryan-mcneil) [reply config/initializers/kms_encrypted_patch.rb:5]: ? My intention was to not have args to this method, so that `**kms_options` isn't required for future models. 

The issue with this, though, is that a couple models DO implement version/previous version, which I've now removed... I'll need to speak to the owner of this file to confirm that's not a problem.

Are you suggesting that we should have optional args for this, so that they _can_ add args, that we ultimately ignore when prepending our version to it?

Comment (ericboehs) [reply config/initializers/kms_encrypted_patch.rb:5]: Maybe we can leave it to a future dev to figure this out. Haha.

But my thought was that if kms_options was nil, we'd set it, otherwise, we'd set it to what we thought it should be. We could potentially merge it as well.

Comment (ryan-mcneil) [reply config/initializers/kms_encrypted_patch.rb:5]: ok, I'm gonna go with current implementation, considering I've confirmed with the codeowner of the form5655 files that this is fine. If someone down the road feels the need to version their records differently than the rest of the app, they can come up with a solution 😄

**Review by kjsuarez — DISMISSED** [review]
LGTM

File: app/sidekiq/kms_key_rotation/batch_initiator_job.rb
Comment (ericboehs) [inline app/sidekiq/kms_key_rotation/batch_initiator_job.rb:41]: Shouldn't this use `KmsEncryptedModelPatch.kms_version` here?

File: app/sidekiq/kms_key_rotation/batch_initiator_job.rb
Comment (ericboehs) [inline app/sidekiq/kms_key_rotation/batch_initiator_job.rb:61]: Do we want to capture the errors and let it continue or should we re-raise here?

Comment (ryan-mcneil) [reply app/sidekiq/kms_key_rotation/batch_initiator_job.rb:62]: so I guess my original thought was that it would log the error, batched_gids would be empty, and the job would then end. But yes, besides the error message in the logs, nothing would tell the system that the job failed and stopped propagating... I'll add something

File: app/sidekiq/kms_key_rotation/batch_initiator_job.rb
Comment (ericboehs) [inline app/sidekiq/kms_key_rotation/batch_initiator_job.rb:58]: Why do we need to encode this into JSON rather than just passing a hash to Sidekiq?

Comment (ericboehs) [reply app/sidekiq/kms_key_rotation/batch_initiator_job.rb:58]: Could we just do:


```ruby
def batched_gids
  records.map(&:to_global_id).each_slice(2).to_a
end
```

And then for the job

```ruby
    def perform(gids)
      Rails.logger.info { "Re-encrypting records: #{gids}" }
      records = GlobalID::Locator.locate_many gids

      skip_hq_callback do
        records.each(&:rotate_kms_key!)
      end
    end
```

Comment (ryan-mcneil) [reply app/sidekiq/kms_key_rotation/batch_initiator_job.rb]: I was able to remove the JSON encoding, but still had to pass the hash, as Sidekiq::Batch doesn't seem to love passing arrays of arrays as args

File: app/sidekiq/kms_key_rotation/batch_initiator_job.rb
Comment (ericboehs) [inline app/sidekiq/kms_key_rotation/batch_initiator_job.rb:43]: Nit: I probably wouldn't include the whole module just for this one method call. Could do `KmsEncryptedModelPatch.kms_version` instead.

Up to you if you want to change. nbd

File: app/sidekiq/kms_key_rotation/batch_initiator_job.rb
Comment (ericboehs) [inline app/sidekiq/kms_key_rotation/batch_initiator_job.rb:46]: Is this needed since we're re-raising? The e.message is already going to be displayed and the stack trace will indicated that it was querying the records.

File: app/sidekiq/kms_key_rotation/batch_initiator_job.rb
Comment (ericboehs) [inline app/sidekiq/kms_key_rotation/batch_initiator_job.rb:56]: Same question as above. Why rescue here?

File: app/sidekiq/kms_key_rotation/rotate_keys_job.rb
Comment (ericboehs) [inline app/sidekiq/kms_key_rotation/rotate_keys_job.rb:17]: ☝️

File: app/models/form5655_submission.rb
Comment (ryan-mcneil) [inline app/models/form5655_submission.rb:13]: discussed with appropriate team (@kjsuarez )

File: lib/periodic_jobs.rb
Comment (ericboehs) [inline lib/periodic_jobs.rb:154]:
```suggestion
  # Rotates Lockbox/KMS record keys and ciphertexts every October 12th (when the KMS key auto-rotate)
```

File: config/initializers/kms_encrypted_patch.rb
Comment (ericboehs) [inline config/initializers/kms_encrypted_patch.rb:12]: It might be good to do some timecop tests around Oct 11th and 12th.

File: app/sidekiq/kms_key_rotation/rotate_keys_job.rb
Comment (ryan-mcneil) [reply app/sidekiq/kms_key_rotation/rotate_keys_job.rb:17]: @ericboehs so I did this one so that we'd have the ID of the record throwing the error. That could be useful, I think

File: lib/periodic_jobs.rb
Comment (ericboehs) [inline lib/periodic_jobs.rb:155]: I wonder if we should start this later in the day? This is set to run at 9pm Pacific.

**Review by ericboehs — APPROVED** [review]

**Pull request merged**
